scalaplugin.py

```diff
@@ -0,0 +1,259 @@
+
+
+def ivy_cache_dir(path):
+    return str(path / "ivycache")
+
+
+def coursier_bin(install_dir):
+    return "{}/coursier".format(install_dir)
+
+
+def bloop_home(install_dir):
+    """
+    The directory bloop thinks is $HOME
+    """
+    return "{}/bloop_home".format(install_dir)
+
+
+def ivy_deps_file(directory):
+    return "{}/ivydependencies.json".format(directory)
+
+
+def scala_version_dep(version):
+    return "org.scala-lang:scala-compiler:{}".format(version)
+
+
+def get_bloop_artifacts():
+    version = "2.12.8"
+    # The version of bsp4s that bloop depeneds on isn't published so we override it
+    deps = ["ch.epfl.scala::bloop-frontend:1.2.5", "ch.epfl.scala::bsp4s:2.0.0-M3"]
+    expanded = [expand_scala_dep(version, dep) for dep in deps]
+    allDeps = expanded + [scala_version_dep(version)]
+    return allDeps
+
+
+def bloop_classpath(coursier, cache, offline=True):
+    deps = get_bloop_artifacts()
+    offlineArgs = ["-m", "offline"] if offline else []
+    cmd = [coursier, "fetch"] + offlineArgs + ["--classpath", "--cache", cache] + deps
+    proc = subprocess.run(cmd, stdout=subprocess.PIPE, universal_newlines=True)
+    return proc.stdout.rstrip()
+
+
+def run_bloop(coursier, bloop_home, cache, args):
+    classpath = bloop_classpath(coursier, cache)
+    if classpath is None:
+        return 1
+    arglist = args.split()
+    set_home = "-Duser.home={}".format(bloop_home)
+    cmd = ["java", "-Xss8M", set_home, "-cp", classpath, "bloop.Cli"] + arglist
+    proc = subprocess.run(cmd)
+    return proc.returncode == 0
+
+
+def fetch_scala_compiler_bridge(coursier, bloop_home, cache, version):
+    # We need absolute paths here
+    cache = os.path.abspath(cache)
+    bloop_home = os.path.abspath(bloop_home)
+    classpath = bloop_classpath(coursier, cache, offline=False)
+    if classpath is None:
+        return 1
+    fetcher = "{}/scala-bridge-fetcher_2.12-0.1.0.jar".format(SCRIPT_DIR)
+    classpath = classpath + ":" + fetcher
+    # Make sure bloop_home exists
+    os.makedirs(bloop_home, mode=0o755, exist_ok=True)
+    set_home = "-Duser.home={}".format(bloop_home)
+    cmd = ["java", set_home, "-cp", classpath, "sifive.ScalaCompilerBridgeFetcher", version]
+    # This creates a target directory, put it in bloop_home
+    proc = subprocess.Popen(cmd, cwd=bloop_home)
+    proc.wait()
+    return proc.returncode == 0
+
+
+def calc_sha256(filename):
+    import hashlib
+    block_size = 65536
+    hasher = hashlib.sha256()
+    with open(filename, 'rb') as f:
+        buf = f.read(block_size)
+        while len(buf) > 0:
+            hasher.update(buf)
+            buf = f.read(block_size)
+    return hasher.hexdigest()
+
+
+def install_coursier(install_dir, jar=False):
+    release_host = "https://github.com/coursier/coursier/releases/download"
+    version = "v2.0.0-RC4-1"
+
+    platform = sys.platform
+    # https://docs.python.org/3.5/library/platform.html#platform.architecture
+    is_64bit = sys.maxsize > 2**32
```

Comment:
Would we really ever run this on a 32bit machine?

Reply:
You'd be surprised what people in the open-source world do. In any case it's trivial to check so might as well be complete.